docking/relative_docking.py

```diff
@@ -0,0 +1,326 @@
+    core (np.array[[int, int], [int, int], ...]): the common core atoms between mol_a and mol_b
+
+    Returns
+    -------
+
+    {str: float}: map of leg label to work values of switching mol_a to mol_b in that leg,
+                  {'protein': [work values], 'solvent': [work_values]}
+
+    Output
+    ------
+
+    stdout noting the step number, lambda value, and energy at various steps
+    stdout noting the work of transition, if applicable
+    stdout noting how long it took to run
+
+    Note
+    ----
+    The work will not be calculated if any norm of force per atom exceeds 20000 kJ/(mol*nm)
+       [MAX_NORM_FORCE defined in docking/report.py]
+    """
+
+    print(
+        f"""
+    HOST_PDBFILE = {host_pdbfile}
+    MAX_LAMBDA = {MAX_LAMBDA}
+    MIN_LAMBDA = {MIN_LAMBDA}
+    TRANSITION_STEPS = {TRANSITION_STEPS}
+    NUM_SWITCHES = {NUM_SWITCHES}
+    """
+    )
+
+    # Prepare host
+    # TODO: handle extra (non-transitioning) guests?
+    print("Solvating host...")
+    (
+        solvated_host_system,
+        solvated_host_coords,
+        _,
+        _,
+        host_box,
+        solvated_topology,
+    ) = builders.build_protein_system(host_pdbfile)
+
+    # sometimes water boxes are sad. Should be minimized first; this is a workaround
+    host_box += np.eye(3) * 0.1
+
+    # Prepare water box
+    print("Generating water box...")
+    # TODO: water box probably doesn't need to be this big
+    box_lengths = host_box[np.diag_indices(3)]
+    water_box_width = min(box_lengths)
+    (
+        water_system,
+        water_coords,
+        water_box,
+        water_topology,
+    ) = builders.build_water_system(water_box_width)
+
+    # sometimes water boxes are sad. should be minimized first; this is a workaround
+    water_box += np.eye(3) * 0.1
+    # it's okay if the water box here and the solvated protein box don't align -- they have PBCs
+
+    # Run the procedure
+    start_time = time.time()
+    guest_name_a = mol_a.GetProp("_Name")
+    guest_name_b = mol_b.GetProp("_Name")
+    combined_name = guest_name_a + "-->" + guest_name_b
+
+    guest_conformer_a = mol_a.GetConformer(0)
+    orig_guest_coords_a = np.array(guest_conformer_a.GetPositions(), dtype=np.float64)
+    orig_guest_coords_a = orig_guest_coords_a / 10  # convert to md_units
+
+    guest_ff_handlers = deserialize_handlers(
+        open(
+            os.path.join(
+                os.path.dirname(os.path.abspath(__file__)),
+                "..",
+                "ff/params/smirnoff_1_1_0_ccc.py",
+            )
+        ).read()
+    )
+    ff = Forcefield(guest_ff_handlers)
+
+    all_works = {}
+    for system, coords, box, label in zip(
+        [solvated_host_system, water_system],
+        [solvated_host_coords, water_coords],
+        [host_box, water_box],
+        ["protein", "solvent"],
+    ):
+        # minimize w.r.t. both mol_a and mol_b?
+        min_coords = minimizer.minimize_host_4d([mol_a], system, coords, ff, box)
```

Comment:
I think it'll be interesting to revisit this point later!

Rather than running an equilibrium simulation at lambda=0 and spinning off 0→1 noneq trajectories, could run an equilibrium simulation at lambda=0.5 and spin off 0.5→1 and 0.5→0 trajectories.

(This could be favorable in some cases, if the 0.5→1, 0.5→0 edges are each at least 2x easier to estimate than the 0->1 edge directly. However, this may not be favorable if it's much more expensive to generate decorrelated equilibrium snapshots at lambda=0.5 than at the lambda=0,1 endstates (e.g. if we have pre-computed a bunch of endstate samples, or if mixing is much slower at lambda=0.5 state).)